### ISSUE TITLE
Support self-message capability with ZNC

### ISSUE THREAD

**Issue by jgeboski** [hexchat/hexchat#1306]
As of znc-1.6.0, a new capability has been introduced, [self-message](https://github.com/znc/znc/commit/382ce76dedb9e7544c38055e1f3752e8ecc46537).  This allows for messages sent from the ZNC user to be seen in other clients connected as the same ZNC user.  This capability should be support just like other ZNC capabilities (`server-time` and `server-time-iso`).

**Comment by TingPing** [comment]
Some work done in https://github.com/hexchat/hexchat/commits/wip/self-message but it's a messy area.
Until then users can use https://github.com/TingPing/plugins/blob/master/HexChat/mymsg.py.

**Comment by Mikaela** [comment]
It should probably also be noted that ZNC calls it as `znc.in/self-message`. I am not fully sure what is `znc.in/echo-message`, but it appears on the list too.

The full list of capabilities that ZNC (1.7.x-git-95-caf6347) offers is: `multi-prefix userhost-in-names znc.in/batch znc.in/echo-message znc.in/playback znc.in/self-message znc.in/server-time-iso`. (znc.in/playback is https://github.com/hexchat/hexchat/issues/1109)

**Comment by TingPing** [comment]
@Mikaela External module, so unrelated.

**Comment by vrcod** [comment]
+1

**Comment by TingPing** [comment]
Last I heard the ircv3 spec removed this. Not sure what zncs plans are.

**Comment by Mikaela** [comment]
It was deprecated in favour of [echo-message](http://ircv3.net/specs/extensions/echo-message-3.2.html) which ZNC also uses.

**Comment by vrcod** [comment]
So if I have the latest version of ZNC, do I have to do anything else on hexchat to make this work? Some settings or a plugin?

**Comment by Mikaela** [comment]
Sorry, I should have been more specific reading my first comment now.

`znc.in/echo-message` was in \*playback module, but is removed from there, because ZNC core now supports `echo-message`. I am not sure if this is in any released version as I am using the `master` branch of ZNC.

Related ZNC issues/prs:
* https://github.com/znc/znc/pull/731
* https://github.com/znc/znc/pull/950
* https://github.com/znc/znc/pull/951

**Comment by vrcod** [comment]
So I'm guessing this is the commit "Implement IRCv3.2 echo-message capability on the "client side" #950" that enables the functionality? If so, then it was only merged 26 days ago and the latest release znc-1.6.0 was on 12th Feb, which means it wouldn't have it yet?

**Comment by TingPing** [comment]
> So if I have the latest version of ZNC, do I have to do anything else on hexchat to make this work? Some settings or a plugin?

Use the [mymsg.py](https://github.com/TingPing/plugins/blob/master/HexChat/mymsg.py) script i wrote.

**Comment by vrcod** [comment]
@TingPing Thanks, I'm using it already, however it seems that self messages only sync up when the client is active... Example: if I close hexchat on my laptop and type a private msg on my desktop hexchat, then reopen hexchat on the laptop - the message I just typed is not there. However if I keep both clients open, they sync up in real time.

**Comment by TingPing** [comment]
@vrcod That is up to if ZNC is configured to auto clear seen buffers or not.

**Comment by vrcod** [comment]
@TingPing I've tried in ZNC with both `AutoClearQueryBuffer=false` (should be the right one - http://wiki.znc.in/Query_buffers) and `AutoClearQueryBuffer=true`, still no change.

**Comment by TingPing** [comment]
worksforme

**Comment by dequis** [comment]
Note that even though self-message was removed from ircv3, it should still be supported without a CAP. See the [removal commit message](https://github.com/ircv3/ircv3-specifications/commit/1bfba47843c2526707c902034b3395af934713c8) for more information.

**Comment by Mikaela** [comment]
Based on previous comments1 I think this issue should be renamed to "Support **echo**-message capability".

**Comment by Mikaela** [comment]
I repeat my previous comment of renaming this issue into "Support echo-message capability" and add that the issue of `server-time` timestamps looking weird when local clock is on wrong time is solved by echo-message.

Today it was reported that at least one user has disabled server-time on #znc because of this issue.

See also:
* https://github.com/ircv3/ircv3-specifications/issues/182
* https://github.com/znc/znc/issues/1144

**Comment by dequis** [comment]
@TingPing What was the issue with that wip branch? I just tried it and (after fixing a small merge conflict in the code that lists supported caps) it works decently enough. [Diff][1]

[1]: https://github.com/hexchat/hexchat/compare/master...dequis:wip/self-message

I don't know enough about the code to review this, though. Only thing that stands out from the diff is `/* There is no Action Send event, so we will just show it as CTCP.. */`

**Comment by TingPing** [comment]
I mentioned on IRC but my main issues are with this commit: https://github.com/hexchat/hexchat/commit/c5d8f59c735e630b357bc2908a3ab4b28d5099ae

I encountered some weird issues with it way back when I tested this and needed to iron those out.

RE Comments on echo-message: I have no interest in echo message (for now at least) and I hope ZNC keeps support for self-message.

**Comment by Mikaela** [comment]
`echo-message` is not ZNC-specific and I think you will get many questions on https://github.com/ircv3/ircv3-specifications/issues/182 when InspIRCd and Charybdis make stable release with `echo-message` support and networks upgrade their IRCds.

**Comment by TingPing** [comment]
For the most part `echo-message` isn't very user facing unless we make UI modifications to show when a message has been received (unlikely to happen).

**Comment by dequis** [comment]
I don't think clients should implement echo-message at all until labeled replies are a thing. There's no reliable way to detect and remove echoes without that (you can't match on message contents since servers may alter the message), other then removing local echo, which leads to poor UX. And that's no trivial matter, clients that implement echo-message that way will seem worse (laggy, "bloated") to users than clients that don't.

But either way, echo-message is a superset of self-message and the implementation of echo-message requires exactly what's in the wip self-message branch, plus some extra stuff.

So let's not talk about that.

**Comment by Arnavion** [comment]
>There's no reliable way to detect and remove echoes without that (you can't match on message contents since servers may alter the message)

IRC is TCP, so messages to a channel are going to be acked in the order they're sent. It would matter if echo-message allowed the server to drop a message without acknowledgement, say if it wanted to supress it completely, but I don't think it does?

>other then removing local echo

Not necessarily. It could appear in a slightly translucent color, and become opaque when received from the server (and modified accordingly). Or it could have a Skype-style spinner next to it to imply it's not yet acknowledged.

**Comment by TingPing** [comment]
@Arnavion Well it probably shouldn't silently drop any but the spec does say:

>  If servers apply any modifications to these messages, they MUST send the final version of the message back to the originating client.

Which makes things harder to track.

> Not necessarily. It could appear in a slightly translucent color, and become opaque when received from the server (and modified accordingly). Or it could have a Skype-style spinner next to it to imply it's not yet acknowledged.

But lets be honest, nobody here is going to do that.

**Comment by dequis** [comment]
The server might also introduce new messages sent from other clients (which is exactly what this ticket is about) so you can't make assumptions about all of them arriving in the order you sent them.

**Comment by Arnavion** [comment]
Right, TingPing explained to me that my assumption of one-message-out-one-message-in doesn't work with bouncers sending own messages from other clients than the current one.

**Comment by ChickenNuggers** [comment]
> But lets be honest, nobody here is going to do that.
IRCCloud does it. Just saying.

**Comment by Arnavion** [comment]
The "nobody here" referred to us HC devs.

**Comment by GaryMatthews** [comment]
is this a bug still? if i use multiple clients and somebody else sends a query message i see that in all connected clients however if i send a message it only goes back to the client that is connected so all other clients appear to have half a conversation. this is not consistent since all connected clients should see my own messages too. i cant go from one client to another client since only half the conversation is there.
